### ISSUE TITLE
kube-prometheus integration

### ISSUE THREAD

**Issue by igorcoding** [lensapp/lens#51]
**Describe the bug**
I couldn't find a way for the Lens to use an existing prometheus installation. In other words, we have an installed kube-prometheus stack but no metrics are visible in the app. Metrics appear only after installing another prometheus stack in the lens-metrics namespace (using the button in the settings).

**Expected behavior**
Kontena Lens automatically (or with some extra configuration) should use an existing prometheus installation

**Environment (please complete the following information):**
- Lens Version: 2.1.4
- OS: Linnux
- Installation method (e.g. snap or AppImage in Linux): AppImage

**Comment by chevalier-m** [comment]
I think this is a pretty mandatory feature.
What's funny is I tried the install for Lens' own Prometheus, and every single node-exporter is pending because the port is already used on every single node. By the exact same deployment on another namespace, same for kube-state-metrics etc....
And, I'm still getting the metrics ! Because they're there and Lens' Prometheus is scraping them.

This thing needs to get debunked, it will also represent a real no-go for many users that you have to install workload that "duplicates" a very common in k8s-prometheus installation.

**Comment by miskun** [comment]
We don't have built-in feature for using existing prometheus with Kontena Lens app yet. This feature is coming soon and we suggest you wait for it unless you are K8S pro user. If you are pro, take a look at the manifests we are using to deploy metrics feature. Adjust to suit your needs: https://github.com/kontena/lens/tree/master/app/manifests/metrics

tl;dr; easy mode for configuring existing Prometheus setup is coming soon.

**Comment by nevalla** [comment]
This feature is now out 🚀 
To enable metrics along `kube-prometheus`, go to cluster settings and use `monitoring/prometheus-k8s:9090` as prometheus service address.

To able to see all metrics on Lens dashboard you are required tweak prometheus labelings a bit:
1. To see node metrics properly, please add
```
    - action: replace
      regex: (.*)
      replacement: $1
      sourceLabels:
      - __meta_kubernetes_pod_node_name
      targetLabel: kubernetes_node
```
relabeling to `node-exporter` servicemonitor crd (for example `kubectl edit -n monitoring servicemonitors node-exporter`).

2. to see cluster's pod usage on cluster overview properly, please add `metricsRelabeling` to `kubelet` service monitor (`kubectl edit -n monitoring servicemonitors node-exporter`)
```
metricRelabelings:
    - action: replace
      sourceLabels:
      - node
      targetLabel: instance
```

**Comment by grv231** [comment]
Hey @nevalla, I have tested lens-Prometheus installation on a test cluster, works sweet!

Additionally, our main clusters have custom Prometheus installation (not kube-prometheus). Currently, the service is running like `prometheus-service` on port 8080 in *monitoring* namespace. I guess the way to add it to cluster settings would be `monitoring/prometheus-service:8080` correct?

Additionally, wanted to confirm the second part of adding the prometheus labelings. We have node-exporters running as daemons (not containers) for scaping node metrics. Any clue what would need to change for those configs? I am not sure where the `metricsRelabelings` part should be added. Are there any docs (Kontena lens related) that I can refer to for this?

**Comment by jnummelin** [comment]
Here's some prometheus related docs related to lens: https://github.com/kontena/lens/blob/master/troubleshooting/custom-prometheus.md

**Comment by grv231** [comment]
Hi @jnummelin while they are helpful, they don't fit my use case since I am not using _kube-prometheus_ or _node exporter_ which the above documents suggest. I am using **Node exporter** as a service on the nodes itself, example - https://prometheus.io/docs/guides/node-exporter/. Moreover, the metrics exposed by node exporter are being scraped by Prometheus (running as a container in kube, along with alertmanager)

So far, I have not been able to use the above settings in my setup and Prometheus fails to understand the config (obvious as the documentation for `cluster's pod usage` might not be related to the setup I am using).

I am not sure right now to create this as a new Github issue, as we have already quite a few pending ones. Have been stuck here for a bit trying to make this work

**Comment by antonkad** [comment]
> This feature is now out 🚀
> To enable metrics along `kube-prometheus`, go to cluster settings and use `monitoring/prometheus-k8s:9090` as prometheus service address.
> 
> To able to see all metrics on Lens dashboard you are required tweak prometheus labelings a bit:
> 
> 1. To see node metrics properly, please add
> 
> ```
>     - action: replace
>       regex: (.*)
>       replacement: $1
>       sourceLabels:
>       - __meta_kubernetes_pod_node_name
>       targetLabel: kubernetes_node
> ```
> 
> relabeling to `node-exporter` servicemonitor crd (for example `kubectl edit -n monitoring servicemonitors node-exporter`).
> 
> 1. to see cluster's pod usage on cluster overview properly, please add `metricsRelabeling` to `kubelet` service monitor (`kubectl edit -n monitoring servicemonitors node-exporter`)
> 
> ```
> metricRelabelings:
>     - action: replace
>       sourceLabels:
>       - node
>       targetLabel: instance
> ```

It would be great to be more specific. I've been playing around for 4h, after installing prometheus-operator, trying to figure out wath's need to be modified in the prometheus configuration.

Wath job need to be modified ? Is it in the prometheus config ? Do i place it randomly ? 

My servicemonitor is named prometheus-operator-node-exporter , do i still need to modify it through the kubectl edit command ?

**Comment by nevalla** [comment]
Lens supports now kube-prometheus out of the box and should not require additional configuration.

**Comment by MRostanski** [comment]
Still does not work out of the box when using kube-prometheus-stack helm installation.

**Comment by ag1989** [comment]
> Still does not work out of the box when using kube-prometheus-stack helm installation.

Similar issue for me, used kube-prometheus-stack and could not get any metrics

**Comment by herberthobregon** [comment]
> Still does not work out of the box when using kube-prometheus-stack helm installation.

Similar issue for me with microk8s.enable prometheus

**Comment by kfirfer** [comment]
> > Still does not work out of the box when using kube-prometheus-stack helm installation.
> 
> Similar issue for me, used kube-prometheus-stack and could not get any metrics

Same


Tried to override this values in the chart:
```
nodeExporter:
  serviceMonitor:
    metricRelabelings:
      - action: replace
        sourceLabels:
          - node
        targetLabel: instance
    relabelings:
      - action: replace
        regex: (.*)
        replacement: $1
        sourceLabels:
          - __meta_kubernetes_pod_node_name
        targetLabel: kubernetes_node

kubelet:
  serviceMonitor:
    metricRelabelings:
      - action: replace
        sourceLabels:
          - node
        targetLabel: instance
```

But unsuccessfully

**Comment by arnoldfranz** [comment]
Same on Kind, doesn't work. Metrics server deployed but Lens ignores the setting.

**Comment by Supesharisuto** [comment]
Just want to leave a comment on this issue since I ran into it.
Lens version 4.0.8
Prometheus setup : Prometheus Operator
To get metrics to show up in Lens: go to cluster->settings
Under Prometheus drop down choose " Prometheus Operator" NOT "Auto detect"
It works for me after this fix

**Comment by Vanshika07** [comment]
Wanted to share what worked for me Using Helm Chart : prometheus-operator
Lens version 4.0.8
Under Settings Add the following
Prometheus Service Address : [namespace]/prometheus-operated:9090
Prometheus installation method : Helm

With the above you will be able to see only Requests and Limits for CPU and Memory but not Node's usage 

To see Cluster's Pod  usage and node metrics properly 

Changes need to be made in ServiceMonitor CRD 

 1. In 'node-exporter' ServiceMonitor  Add (Under spec: then endpoints:)
	relabelings:
        - action: replace
          regex: (.*)
          replacement: $1
          sourceLabels:
            - __meta_kubernetes_pod_node_name
          targetLabel: kubernetes_node
		  
  2. In 'Kubelet' ServiceMonitor add (Under spec: then endpoints:)
     metricRelabelings:
          - action: replace
            sourceLabels:
            - node
            targetLabel: instance

All metrics are shown perfectly after this fix

**Comment by socar-hardy** [comment]
Same problem here.

I used (installed)

- private gke cluster (1.19.7-gke.1302)
- kube-prometheus-stack 13.13.0
  - (https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack)
- lens 4.1.3

and I added extra configuration following
https://github.com/lensapp/lens/blob/master/troubleshooting/custom-prometheus.md

and all resource appear well

![image](https://user-images.githubusercontent.com/62049738/109113822-2da15880-7780-11eb-832b-ebba6bd99078.png)

and I set lens configuration like this

![image](https://user-images.githubusercontent.com/62049738/109113875-41e55580-7780-11eb-926c-989d9e60c5b8.png)

but still metric didn't appear.... (infinitely loading)

![image](https://user-images.githubusercontent.com/62049738/109113942-59244300-7780-11eb-9fa3-efa010a47041.png)

How can I solve this problem..?!

**Comment by jakolehm** [comment]
@socar-hardy probably related to a private GKE, see discussion here: https://github.com/lensapp/lens/issues/1731#issuecomment-779196430

**Comment by jgrau** [comment]
I was experiencing the same issue with `kube-prometheus-stack` but got it working today. I believe my issue was that I was overriding the namespace in which `kube-state-metrics` was installing (which I did to try to get "Advanced Metrics" in the Digital Ocean UI). Here's me config which is working:

```
# values.prometheus.yaml
kubelet:
  serviceMonitor:
    https: false
    metricRelabelings:
    - action: replace
      sourceLabels:
      - node
      targetLabel: instance

nodeExporter:
  serviceMonitor:
    relabelings:
      - action: replace
        regex: (.*)
        replacement: $1
        sourceLabels:
        - __meta_kubernetes_pod_node_name
        targetLabel: kubernetes_node
```

```
# helmfile.yaml (to display the namespace and exact chart)
  - name: kube-prometheus-stack
    namespace: monitoring
    chart: prometheus-community/kube-prometheus-stack
    version: 13.x
    values:
      - values.prometheus.yaml
      - secrets.prometheus.yaml

# helmfile.lock (just to display the helm release version
- name: kube-prometheus-stack
  repository: https://prometheus-community.github.io/helm-charts
  version: 13.13.1
```

Lens config:

<img width="2025" alt="Screenshot 2021-06-02 at 09 49 42" src="https://user-images.githubusercontent.com/108088/120444392-de268600-c387-11eb-8c34-fe83f2ecb86d.png">

Setup:

Kubernetes on Digital Ocean
Kube Prometheus Stack

**Comment by jgrau** [comment]
Ehh. UPDATE: I was getting cluster level metrics with the above configuration but not pod level metrics. Turns out removing the custom config in `values.yaml` resultet in pod level metrics as well. I believe that means that for me, leaving all settings alone works and my problem was trying to override the kube-state-metrics namespace. I could even set Lens to Auto-Detect my prometheus installation and it still worked.

**Comment by gjensen-gomeyra** [comment]
I used the `loki-stack` helm chart from https://grafana.github.io/loki/charts to setup a full PLG stack.  Here is my final configuration that works with all Lens metrics:

```yaml
grafana:
  enabled: true
  persistence:
    enabled: true
    storageClassName: default
prometheus:
  enabled: true
  alertmanager:
    persistentVolume:
      storageClass: default
  server:
    persistentVolume:
      storageClass: default
    global:
      scrape_interval: 30s
  kubelet:
    serviceMonitor:
      relabelings:
      - action: replace
        regex: (.*)
        replacement: $1
        sourceLabels:
        - __meta_kubernetes_pod_node_name
        targetLabel: kubernetes_node
      metricRelabelings:
      - action: replace
        sourceLabels:
        - node
        targetLabel: instance
```

**Comment by mattduguid** [comment]
We had a requirement to get this working via terraform for our AKS clusters and were successful using 2 steps,

**step 1 - terraform**

resource "helm_release" "monitoring" {
   atomic = true
   chart = "kube-prometheus-stack"
   name = "kube-prometheus-stack"
   create_namespace = true
   namespace = "lens-metrics"
   repository = "https://prometheus-community.github.io/helm-charts/"
   timeout = 180 
   wait = true 
   set {
     name  = "rbacEnable"
     value = "false"
   }
   depends_on = [
     azurerm_kubernetes_cluster.azure,
   ]
}

**step 2 - change cluster setting in lens IDE**

In the lens IDE we also needed the extra bit mentioned by @Supesharisuto which was,

_To get metrics to show up in Lens: go to cluster -> settings Under Prometheus drop down choose " Prometheus Operator" NOT "Auto detect"_

![image](https://user-images.githubusercontent.com/7642305/128400845-1c5c4330-b85a-4547-bb3a-627a3056c83e.png)

After that metrics started showing

**Comment by joriskoris-emburse** [comment]
Unable to get metrics working here either.

I have a basic `kube-prometheus` setup on EKS, followed https://github.com/lensapp/lens/blob/master/troubleshooting/custom-prometheus.md as well.

How does lens query metrics behind the scenes? I assume it's via proxies (https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-services/#manually-constructing-apiserver-proxy-urls)?

**Comment by j-rewerts** [comment]
I was having issues as well and was able to get mine working.

To investigate, check out [this file](https://github.com/lensapp/lens/blob/master/src/main/prometheus/lens.ts) for the queries they're using. Note that's the default "Lens" provider, but you may be using the Helm or Operator providers. Just check the other files to be sure. If you want to see how they set up their built-in prometheus feature, check out [this configmap](https://github.com/lensapp/lens/blob/master/extensions/metrics-cluster-feature/resources/02-configmap.yml.hb). In my case, I needed to add a little snippet to my node-exporter job.
``` yaml
- action: replace
  source_labels:
  - __meta_kubernetes_pod_node_name
  target_label: kubernetes_node
```

**Comment by ndrscodes** [comment]
I'm still having this issue using kube-prometheus-stack. After editing the node-exporter ServiceMonitor CRD, the pod-level statistics work flawlessly. Sadly, no matter what i do, i never seem to get the node/cluster level statistics to display anything. It always shows the "Metrics are not availabe...." error message.
Interestingly, auto-detect works in my case (like it did for @jgrau)

**Comment by ndrscodes** [comment]
Update: It works using Lens 5.2.7. Additionally, i had to move from using kube-prometheus-stack to using the prometheus operator. Reporting works now. Only issue is that the CPU Usage is displayed as 0.00 in the gauge. In the graph, it is reported to be 0.8.

**Comment by Jeansen** [comment]
Current versions 5.3.3 works for me with Prometheus 4.12.0 helm chart and setting the metrics detection to helm in cluster settings. Auto detection did not work! Curiously, Using Prometheus 15.x.x did not work for me. The same is true with current versions of kube-prometheus-stack. Grafana boards are filled. I also can see metrics in `kubectl top nodes` or `kubectl top pod`, but not in the Lens app. So, I'll stick with Prometheus 4.12.0 for the moment. My cluster is purely educational, anyway.

**Comment by ndrscodes** [comment]
seems like this was fixed in #4617. Hopefully everything will work in the next version :)

**Comment by ypicard** [comment]
Prometheus 5 metrics are not displayed in Lens 5.4.6 for me.

**Comment by Jeansen** [comment]
What worked partly for me was some part of https://github.com/lensapp/lens/issues/4435#issuecomment-1129179399

I put the following in a custom values.yaml:

```
prometheus:
  monitor:
    enabled: false
    metricRelabelings:
      - action: replace
        regex: (.*)
        replacement: $1
        sourceLabels:
          - __meta_kubernetes_pod_node_name
        targetLabel: kubernetes_node
```

Then I installed the metrics-server via prometheus chart:

```
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update
helm install  kube-metrics prometheus-community/kube-state-metrics -f values.yaml
```

After that, I installed prometheus:

```
helm repo addrometheus-community https://prometheus-community.github.io/helm-charts/
helm repo update
helm install --set server.global.scrape_interval=30srometheusrometheus-community/prometheus --version=15.8.7'"
```

But with this I only see the memory usage on nodes, no disk or cpu. Also, I have to tell Lens that prometheus metrics have been installed with helm. Neither the operator or auto detection settings work. Better than nothing, but not perfect. Ultimately, I reverted back to prometheus 4.12.0 because that still works out of the box.
